### PR TITLE
Update Rust to 1.83, and fix clippy warnings from that version

### PR DESCRIPTION
**Pull request by mhammond**
### Pull Request checklist ###

- **Breaking changes**:  This PR follows our [breaking change policy](https://github.com/mozilla/application-services/blob/main/docs/howtos/breaking-changes.md)
  - [ ] This PR follows the breaking change policy:
     - This PR has no breaking API changes, or
     - There are corresponding PRs for our consumer applications that resolve the breaking changes and have been approved
- [ ] **Quality**: This PR builds and tests run cleanly
  - Note:
    - For changes that need extra cross-platform testing, consider adding `[ci full]` to the PR title.
    - If this pull request includes a breaking change, consider [cutting a new release](https://github.com/mozilla/application-services/blob/main/docs/howtos/cut-a-new-release.md) after merging.
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes a changelog entry in [CHANGELOG.md](../CHANGELOG.md) or an explanation of why it does not need one
  - Any breaking changes to Swift or Kotlin binding APIs are noted explicitly
- [ ] **Dependencies**: This PR follows our [dependency management guidelines](https://github.com/mozilla/application-services/blob/main/docs/dependency-management.md)
  - Any new dependencies are accompanied by a summary of the due diligence applied in selecting them.

[Branch builds](https://github.com/mozilla/application-services/blob/main/docs/howtos/branch-builds.md): add `[firefox-android: branch-name]` to the PR title.

### CODE REVIEW

File: components/as-ohttp-client/Cargo.toml [hunk 1]
```diff
@@ -25,3 +25,7 @@ features = ["client", "server", "app-svc", "external-sqlite"]
 
 [build-dependencies]
 uniffi = { workspace = true, features=["build"]}
+
+# Temp hack for uniffi 0.28 generated code, will not be necessary after next uniffi update.
+[lints.clippy]
+empty-line-after-doc-comments = "allow"
```
Comment (mhammond) [inline components/as-ohttp-client/Cargo.toml:31]: This is very unfortunate :( We could put this in the top-level Cargo.toml, until we get to m-c, where we will be unable to use workspace attributes for almost all of these crates.

Comment (linabutler) [reply components/as-ohttp-client/Cargo.toml:31]: I think this is totally OK for now—as you said in the comment, it's temporary!

**Comment by linabutler** [comment]
AFAIK, m-c is [still on 1.76.0](https://searchfox.org/mozilla-central/rev/db4c501abb679a81c921510d7e38a8028dd6024a/python/mozboot/mozboot/util.py#15)—are all the 1.83 changes backward-compatible, or should we see about bumping m-c's version to match?

**Review by linabutler — APPROVED** [review]
Clippy is getting better all the time (I didn't realize that eliding lifetimes in `impl` headers has been possible [since 1.31](https://doc.rust-lang.org/stable/releases.html#version-1310-2018-12-06), for example!). Cool! 🚢

**Comment by mhammond** [comment]
> AFAIK, m-c is [still on 1.76.0](https://searchfox.org/mozilla-central/rev/db4c501abb679a81c921510d7e38a8028dd6024a/python/mozboot/mozboot/util.py#15)—are all the 1.83 changes backward-compatible, or should we see about bumping m-c's version to match?

IIUC, that's the min version, but all our actual builders are on this version - https://bugzilla.mozilla.org/show_bug.cgi?id=1909343

**Comment by mhammond** [comment]
not sure what the iOS failure is all about though, but that's tomorrow's problem!

**Comment by linabutler** [comment]
I reran the workflow on a hunch, and now it's green 🤷🏼‍♀️ I'm a little surprised that it timed out after 5 seconds, and I wonder if there's something deeper for us to look at there—but it doesn't look like it's related to the MSRV bump.

**Pull request merged**